### PR TITLE
nixos: add python testing support

### PR DESCRIPTION
**Pull request by tfc**
##### Motivation for this change

Myself and most people i work with don't know Perl but Python. NixOS integration tests are totally awesome and more people should use/write such to test software.

I would like to support the acceptance and adoption of NixOS integration tests by letting users choose the more popular language Python to implement the imperative integration test part.

###### Things done

I ported the test driver to python and forked the nix scripts. Of course the duplication can be reduced, but i did not want to put more effort into this in case this work is not accepted in this repo for some reason, so i am eager to hear/read your thoughts.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @edolstra @thefloweringash @abbradar @samueldr @srhb 

# Community discussion

After discussing this with some people at nixcon (including @edolstra @Lassulus @adisbladis @domenkozar @garbas ...) i arrived at the following TODO-list to make the whole thing merge-acceptable (Note that there was *no decision*):

- [ ] finish porting all funcitonality from the perl driver to python (ocr, coverage, ... it's not much missing)
- [ ] port *all* tests to python (so far these are roughly ~300 tests)
- [ ] *remove* perl test driver so there is only one
- [ ] squash all commits with good commit messages (following @jonringer's suggestion)

So far @blitz @jtraue @mdvjd are helping me already

### CODE REVIEW

**Comment by jonringer** [comment]
Do you mind changing the title and your first commiit to something like `nixos: add python testing support`

the later 3 commits according to contributing.md should be:
```
nixos/bittorent: convert test to python
nixos/bittorent: convert test to python

**Comment by jonringer** [comment]
but this is awesome, i think python will enable a lower barrier-to-entry for nixos than Perl.

**Comment by tfc** [comment]
@jonringer I fixed the commit messages according to your suggestions.

**Comment by jonringer** [comment]
seeing as python support might be a disruptive change, please allow for some time for the necessary communication to happen :)

**Comment by blitz** [comment]
@tfc Super nice! Life is too short to learn Perl. :)

**Comment by jonringer** [comment]
cc @worldofpeace @FRidh

**Comment by FRidh** [comment]
I haven't written any NixOS tests myself, but is the language (Perl) such a barrier that we need to implement it (also) in Python? From what I understand the tests are often quite simple, not requiring any deep knowledge of Perl. Personally I don't mind having it in Python at all, but let's keep this open for a while so those that use the NixOS testing framework can review.

Also, it seems to be we need a code owner for the original parts and the newly added parts.

**Comment by tfc** [comment]
@FRidh perl is not a barrier to using nixos integration tests at all. Once you get used to them you would like to write fancier and fancier tests that use the language's facilities to reduce duplication and to make things a bit more dynamic, and maybe also add libraries (currently not supported in my proposal yet, but i also don't want to blow up the PR). That is where non-Perl users would like to use Python.

If an owner is needed for the Python part, i would volunteer as this would be heavily used by my colleagues at work.

**Comment by blitz** [comment]
Maybe an interesting data point: https://trends.google.com/trends/explore?date=today%205-y&geo=US&q=%2Fm%2F05zrn

**Comment by worldofpeace** [comment]
TBH, I would play with testing **much** more strongly if it was in python.
I don't think it's too unreasonable to support multiple types of testing scripts.
Can we not have a `testing-python.nix` file? It's the same as `testing.nix` at a glance,  but just changed to use the python things. So `testing.nix` needs support to use different testing drivers.

Though I'm realizing you've done this on purpose so as to get feedback.
Currently short on time, but I can give in depth thoughts and opinions soon.

**Comment by domenkozar** [comment]
cc @copumpkin  you wanted this :D

**Comment by tfc** [comment]
Cc @adisbladis

**Comment by tfc** [comment]
Cc @azazel75

**Comment by jonringer** [comment]
@tfc I would hold off on converting a lot of the tests over to python. It will expand how much this PR is changing and make it harder to come to a decision on whether or not it should be merged. 

Even though a lot of the converted tests seem to be a 1:1 translation :)

**Comment by worldofpeace** [comment]
> After discussing this with some people at nixcon (including @edolstra @Lassulus @adisbladis @domenkozar @garbas ...) i arrived at the following TODO-list to make the whole thing merge-acceptable:

So it appears the people you've talked to at nixcon agreed on porting testing completely to python?
Not that I disagree with that, maintaining multiple testing drivers doesn't really sound all that great.

Can the people who discussed this at nixcon publicize their viewpoints here?
Because currently it seems there was an agreement to port testing completely to python and to accept this change, and maybe not everyone agrees with this, or at least the people @tfc had discussion with agreed with it. Mostly looking out for people so they can know why.

If my assumption here is incorrect, correction is welcome.

**Comment by jonringer** [comment]
oh, i didn't see that he updated the original PR post. Very nice.

It might be a good idea to get this initial addition merged, then have the re-writes separated into a series of PRs so they are easier to test and smaller scope

**Comment by tfc** [comment]
@worldofpeace @jonringer There was *no decision* and no one's opinion was overrided. I talked to many people about their opinion and also wanted to find out what people of the "inner" parts of the community think about how to get this merge-acceptable. There was no promise that this PR will be merged as soon as i ticked all the boxes that i added to the PR post.

On the Nixcon hackday @blitz @jtraue @mdvjd and i ported more things because we were in the mood.

**Comment by worldofpeace** [comment]
Ok, thanks for clearing that up @tfc. And I do agree with those things completed your submission will look very attractive to those who agree that nixos testing should be in python.

**Comment by jonringer** [comment]
I appreciate the work, and I find being able to write tests in python much more attractive than perl :).

Keep fighting the good fight :).

**Comment by azazel75** [comment]
jI too hope that this very good work is merged! Thanks @tfc @blitz @jtraue @mdvjd for you work!

**Comment by flokli** [comment]
Thanks a lot from my side too :-) IMHO, it's much more readable, and seems better refactorable, if we reuse some of the harness if we ever do NixOS Container-based tests (or use lighter hypervisors).

piggy-backing on the existing code, @adisbladis and I thought about forwarding the dbus system bus to the host, to do all systemd related things via `pystemd`
.
This should offer a more native way to interact with systemd instead of parsing command output, and provide a more native way to wait for targets to be reached and inspect unit states.

There's a WIP at https://github.com/flokli/nixpkgs/tree/integration-test-python-pystemd, but I somehow couldn't get the serial to work - if anybody has an idea on how to fix this, highly appreciated ;-)

**Comment by tfc** [comment]
@flokli i had a look at the code. This is cool. I also left you a comment on one of the commits. I am all for such things, but i am not sure if it makes it easier to get the python support in general if we do this at the same time. What do you think?

**Comment by flokli** [comment]
@tfc I didn't meant to stall this PR, this was only meant as a heads-up, and can handled as a follow-up.

File: nixos/lib/test-driver/test-driver.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
+#! /somewhere/python3
+
+from __future__ import print_function
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: This shouldn't be needed on Python 3 - we're already in the future.

File: nixos/lib/test-driver/test-driver.py [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+from xml.sax.saxutils import XMLGenerator
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: Can we split some of the classes into separate files? I was a bit baffled about why we need to generate xml, and the list is quite long.

File: nixos/lib/test-driver/test-driver.py [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+import atexit
+import _thread
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]:
```suggestion
import thread
```

File: nixos/lib/test-driver/test-driver.py [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+    'H': 'shift-h', 'U': 'shift-u', '\\': '0x2B', '|': 'shift-0x2B',
+    'I': 'shift-i', 'V': 'shift-v',  ',': '0x33', '<': 'shift-0x33',
+    'J': 'shift-j', 'W': 'shift-w',  '.': '0x34', '>': 'shift-0x34',
+    'K': 'shift-k', 'X': 'shift-x',  '/': '0x35', '?': 'shift-0x35',
+    'L': 'shift-l', 'Y': 'shift-y',  ' ': 'spc',
+    'M': 'shift-m', 'Z': 'shift-z', '\n': 'ret',
+    '!': 'shift-0x02', '@': 'shift-0x03', '#': 'shift-0x04', '$': 'shift-0x05',
+    '%': 'shift-0x06', '^': 'shift-0x07', '&': 'shift-0x08', '*': 'shift-0x09',
+    '(': 'shift-0x0A', ')': 'shift-0x0B',
+    }
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: did you run this through a linter? Mine complained about the closing bracket not matching indentation of the opening bracket"s line.

In general, we should just always lint before copying in `testing-python.nix`.

File: nixos/lib/test-driver/test-driver.py [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+    for n in range(900):
+        if fn(False):
+            return
+        time.sleep(1)
+
+    if not fn(True):
+        raise Exception('action timed out')
+
+
+class Logger:
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: can we make use of python's native logging, by any chance?

This would avoid mixing the creating of log records with the different representations.

I assume nesting might be hard, but still wanted to ask :-)

File: nixos/lib/test-driver/test-driver.py [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+        else:
+            try:
+                cmd = args['startCommand']
+                self.name = re.search('run-(.+)-vm$', cmd).group(1)
+            except:
+                self.name = 'machine'
+
+        self.script = args.get('startCommand', self.create_startcommand(args))
+
+        tmp_dir = os.environ.get('TMPDIR', '/tmp')
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: can we use `tempfile.tempdir` here instead?

File: nixos/lib/test-driver/test-driver.py [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+
+        self.booted = False
+        self.connected = False
+        self.pid = None
+        self.socket = None
+        self.monitor = None
+        self.logger = args['log']
+        self.allow_reboot = args.get('allowReboot', False)
+
+    def create_startcommand(self, args):
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: this can be a static method, it's not acessing `self` at all.

**Review by flokli — CHANGES_REQUESTED** [review]
I also couldn't get the interactive repl to work:

```
nix-build -A nixos/tests/….nix -A driver
result/bin/nixos-test-driver
```

File: nixos/lib/test-driver/test-driver.py (hunk 1 above)
Comment (tfc) [reply nixos/lib/test-driver/test-driver.py]: Thx for your review! I won't be able to do any of this today, but i will incorporate soon!

File: nixos/lib/test-driver/test-driver.py (hunk 7 above)
Comment (tfc) [reply nixos/lib/test-driver/test-driver.py]: @flokli i incorporated some of the suggestions, but i would like to discuss others. Can you maybe send me a message via matrix/irc? I hang out in the nixos channels with the nick "jonge".

File: nixos/lib/test-driver/test-driver.py (hunk 2 above)
Comment (edolstra) [reply nixos/lib/test-driver/test-driver.py]: I think we can drop the XML log file stuff. This was for generating pretty-printed logs with collapsible structure, but I don't think anybody cares about keeping that.

File: nixos/lib/testing-python.nix [hunk 8]
```diff
@@ -0,0 +1,266 @@
+{ system
```
Comment (edolstra) [inline nixos/lib/testing-python.nix]: Do we need a separate `testing-python.nix` which is mostly identical to `testing.nix`? Maybe we can have a single `makeTest` function that dispatches based on whether the test has a `testScript` or `pythonTestScript` function.

File: nixos/lib/testing-python.nix [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+          LOGFILE=$out/log.xml tests='eval $ENV{testScript}; die $@ if $@;' ${driver}/bin/nixos-test-driver
+
+          # Generate a pretty-printed log.
+          xsltproc --output $out/log.html ${./test-driver/log2html.xsl} $out/log.xml
+          ln -s ${./test-driver/logfile.css} $out/logfile.css
+          ln -s ${./test-driver/treebits.js} $out/treebits.js
+          ln -s ${jquery}/js/jquery.min.js $out/
+          ln -s ${jquery}/js/jquery.js $out/
+          ln -s ${jquery-ui}/js/jquery-ui.min.js $out/
+          ln -s ${jquery-ui}/js/jquery-ui.js $out/
```
Comment (edolstra) [inline nixos/lib/testing-python.nix]: This can all be removed if we drop the XML logging.

**Comment by edolstra** [comment]
Looks good to me. I agree with the summary of the NixCon discussion. We should get rid of the perl framework in order not to have to maintain two frameworks.

File: nixos/lib/testing-python.nix (hunk 8 above)
Comment (tfc) [reply nixos/lib/testing-python.nix]: In different cases there would be different outcomes:

- keep two test drivers: reduce this duplication
- keep only python driver: remove this file, make the other work with python.

Community discussion results point towards the latter, so i will remove once it's all ported.

File: nixos/lib/testing-python.nix (hunk 9 above)
Comment (tfc) [reply nixos/lib/testing-python.nix]: I did not yet take into consideration to drop the XML logging - just ported it to work exactly the same in the python version. Are there any constraints for an alternative implementation without xml from how people use/read this logs?

File: nixos/lib/test-driver/test-driver.py [hunk 10]
```diff
@@ -0,0 +1,685 @@
...
+        for machine in machines:
+            if machine.pid is None:
+                continue
+            log.log('killing {} (pid {})'.format(machine.name, machine.pid))
+            machine.process.kill()
+
+    log.close()
+
+
+tic = time.time()
```
Comment (Mic92) [inline nixos/lib/test-driver/test-driver.py]:
```python
if __name__ == "__main__":
  begin = time.time()
  run_tests()
  end = time.time()
  print('test script finished in {:.2f}s'.format(end - begin))
```

**Comment by Mic92** [comment]
Please also update: `doc/manual/development/writing-nixos-tests.xml`

File: nixos/lib/test-driver/test-driver.py (hunk 2 above)
Comment (azazel75) [reply nixos/lib/test-driver/test-driver.py]: @edolstra I think that the xml/html log help to give context to the log entries and to tie them to the imperative command that is the cause for their generation. It does a very good job at it, with its tree-like rendering.  The only thing that i find strange is the face that when the test fails, it isn't generated. It would be much more useful to understand what happened if we could have it when the test fails

File: nixos/lib/test-driver/test-driver.py (hunk 10 above)
Comment (tfc) [reply nixos/lib/test-driver/test-driver.py]: why would you only move those 4 lines into the conditional, while the rest of the script already started all the machines? what is the use case?

File: nixos/lib/testing-python.nix [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+  testDriver = stdenv.mkDerivation {
+    name = "nixos-test-driver";
+
+    buildInputs = [ makeWrapper python3Full ];
+
+    dontUnpack = true;
+
+    preferLocalBuild = true;
+
```
Comment (FRidh) [inline nixos/lib/testing-python.nix]: Could there be a `checkPhase` that runs `pylint --errors-only` and `black --check` to verify logically as much as possible and that we have a common style (without style discussions)
https://github.com/psf/black

Comment (tfc) [reply nixos/lib/testing-python.nix]: oh `black` is nice... this could also be run over *all* test scripts.

**Comment by tfc** [comment]
@flokli @Mic92 i incorporated most of your comments and i suggest handling the others in subsequent pull requests (and maybe discuss details offline, what would be the right medium for that?).

What do you think? Could we maybe merge this already as @edolstra suggested and finish the rest of the tests in incremental steps?

**Comment by flokli** [comment]
regarding discussion: I joined `#nixos-python-test-driver`, happy to do conversations there too.

File: nixos/lib/test-driver/test-driver.py (hunk 10 above)
Comment (Mic92) [reply nixos/lib/test-driver/test-driver.py]: Actually I would put it in a `main()` function, so that later one something like mypy can be used for the driver for type-checking.

File: nixos/lib/testing-python.nix [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+      # VMs from the command line.
+      driver = runCommand testDriverName
+        { buildInputs = [ makeWrapper];
+          testScript = testScript';
+          preferLocalBuild = true;
+          testName = name;
+        }
+        ''
+          mkdir -p $out/bin
+          echo "$testScript" > $out/test-script
```
[the remark this answers is not in the document]
Comment (tfc) [reply nixos/lib/testing-python.nix]: `pylint` doesn't recognize the per-machine variables... but `black` works.

**Review by flokli — COMMENTED** [review]
please splice the `black`-related changes in the corresponding original "change test to python" commits.

File: nixos/lib/testing-python.nix [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+  inherit pkgs;
+
+
+  testDriver = let
+    testDriverScript = ./test-driver/test-driver.py;
+  in stdenv.mkDerivation {
+    name = "nixos-test-driver";
+
+    buildInputs = [ makeWrapper python3Full ]
```
Comment (flokli) [inline nixos/lib/testing-python.nix]: `python3` should be enough.

File: nixos/lib/testing-python.nix [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+
+  testDriver = let
+    testDriverScript = ./test-driver/test-driver.py;
+  in stdenv.mkDerivation {
+    name = "nixos-test-driver";
+
+    nativeBuildInputs = [ makeWrapper ];
+    buildInputs = [ python3 ];
+    checkInputs = with python3Packages; [ pylint black];
```
Comment (flokli) [inline nixos/lib/testing-python.nix]:
```suggestion
    checkInputs = with python3Packages; [ pylint black ];
```

**Review by flokli — CHANGES_REQUESTED** [review]
Please squash the black-related fixes into the the individual python test rewrites, and put the "enabling black support commit" before these.

**Comment by flokli** [comment]
Also, `nixos/doc/manual/development/writing-nixos-tests.xml` should be `make-test-python.nix` - we don't replace `make-test.nix` yet, and following the docs after this PR is merged would just fail.

File: nixos/lib/test-driver/test-driver.py (hunk 10 above)
Comment (domenkozar) [reply nixos/lib/test-driver/test-driver.py]: I second that, being able to use this file as a library has many benefits (including easier understanding when reading the code).

Also gets rid of global state for when [strict python](https://instagram-engineering.com/python-at-scale-strict-modules-c0bb9245c834) becomes a thing :)

File: doc/contributing/submitting-changes.xml [hunk 15]
```diff
@@ -375,29 +375,27 @@ Additional information.
 
   <section xml:id="submitting-changes-master-branch">
    <title>Master branch</title>
+
    <para>
-    The <literal>master</literal> branch is the main development branch. 
-    It should only see non-breaking commits that do not cause mass rebuilds.
+    The <literal>master</literal> branch is the main development branch. It should only see non-breaking commits that do not cause mass rebuilds.
```
Comment (flokli) [inline doc/contributing/submitting-changes.xml]: The changes in `submitting-changes.xml` seem unrelated.

**Review by flokli — APPROVED** [review]
Thanks a lot for this!

**Comment by worldofpeace** [comment]
@flokli Why wasn't https://github.com/NixOS/nixpkgs/pull/71684/files#r340032396 done?
I believe I agreed with @edolstra here https://github.com/NixOS/nixpkgs/pull/71684#issuecomment-545999187. Is that going to be in another PR?

**Comment by flokli** [comment]
@worldofpeace there was no clear consensus on https://github.com/NixOS/nixpkgs/pull/71684#discussion_r340015544, and as this can always be done in a follow-up PR, I didn't want to stall this any further.

**Comment by worldofpeace** [comment]
@flokli I think I linked the wrong comment https://github.com/NixOS/nixpkgs/pull/71684/files#r340031960

**Comment by flokli** [comment]
In that case, porting isn't finished, but I didn't consider this a requirement for this PR, which is why I asked about https://github.com/NixOS/nixpkgs/pull/71684#issuecomment-549483588.

I assume porting the remaining nixos vm tests will happen soon, and once that's done, we can remove the perl test (and maybe rename the python one to the original location).

Docs already only document the python one.

File: nixos/lib/test-driver/test-driver.py (hunk 2 above)
Comment (arianvp) [reply nixos/lib/test-driver/test-driver.py]: That was done on purpose to unblock issues with Hydra IIRC.

**Comment by volth** [comment]
few months ago I got rid rid of `perl` in `test-driver` in favor of `scala` + `monix`, those allow to write multi-VMs tests expressively.

I do not pitch the solution here, as a _barrier_ it is taller than that of `perl`.

But... `python` does not look an improvement at all, it differs from `perl` only in syntax, and the `python` syntax seems weirder to be used in ''-string literals whe Nix strips whitespaces.

**Comment by volth** [comment]
This one is especially nice improvement showing advantage of python whitespace syntax:

https://github.com/nixos/nixpkgs/blob/40396a7e12419fc772063d2a3fa87ec7d14a46d9/nixos/tests/boot.nix#L20-L21

**Comment by FRidh** [comment]
Without going into whether Python is an improvement or not, but in these cases we should just write the data as JSON and read that. That means currently explicitly writing to JSON and in the future with structured attributes we can simply pass the dict as an attribute.

**Comment by domenkozar** [comment]
> But... python does not look an improvement at all, it differs from perl only in syntax, 

That's untrue, the major difference is that nowadays Python is on the top of the list of popular languages, so probably millions more are capable of contributing easily. 

> Even nodejs seems favorable than python.

That could be argued, but it's least designed language of the popular ones we have.

**Comment by volth** [comment]
> > But... python does not look an improvement at all, it differs from perl only in syntax,
> 
> That's untrue, the major difference is that nowadays Python is on the top of the list of popular languages, so probably millions more are capable of contributing easily.
> 

I see a high potential here developing this argument: could we replace `stdenv`'s `bash` with something else?
No jokes, I am serious, and will be happy even with `python`

**Comment by FRidh** [comment]
@volth if you would like to discuss the future of the stdenv (language), please open a separate issue. I am sure others (including myself) would be interested in discussing it.

**Comment by volth** [comment]
@FRidh it is open somewhere, in both nixpkgs and nix.
But I see it is somehow related.

Unlike Guix which used Scheme for everywhere, Nix lacks own imperative language. Or better to say it just got `python` as the 3rd embedded imperative language in addition to `bash` and `perl`.

**Comment by domenkozar** [comment]
Quite a few scripts have been rewritten to python, so I'd rather say: we're replacing `perl` for `python`.

**Comment by domenkozar** [comment]
```
$ git co release-16.03
$ find . -name "*.py" | wc -l
1

$ find . -name "*.pl" | wc -l
8
```

```
git co master
$ find . -name "*.pl" | wc -l
7

$ find . -name "*.py" | wc -l
5
```

Where we have yet to remove `test-driver.pl` some day.

**Comment by volth** [comment]
> Quite a few scripts have been rewritten to python, so I'd rather say: we're replacing `perl` for `python`.

It is exactly the point of critic.

Could we consider a vision of a single imperative language for the whole Nix ecosystem and  replace `perl` with it and then `bash` and other places where imperative code is used or (may be used in the future) ?

It could happen that it will be `python` at the end.
What I do not like is what the election has been done from a single candidate and also because `python` has some weaknesses (whitespacing syntax, inexpressive async code, ...) which have not been considered (I also recall @edolstra post about "it was a mistake to write nixops in `python`").

**Comment by volth** [comment]
> > scala + monix,
> 
> Well scala is used in very specific settings, so that way I'd actually prefer perl since it's more approachable.
> I think you're forgetting that any NixOS developer should be able to write these tests.

No, I do not forget, my next sentence was "I do not pitch my solution here, as a barrier it is taller than that of perl".
My point of replacement `perl`->`scala` to have support for async, which is very useful for `test-driver`.
`python` has some support for async too, although it was not articulated as the reason of `perl`->`python` transition.
So, I would call async support of the possible `perl` successor as a requirement, and `python` here is on the list (among almost all nodejs and jvm languages)

**Comment by aszlig** [comment]
Maybe it's a bit late, but just wanted to throw in my two cents on this (since I'm buried in other things lately :-/): I do like the switch, since I'm already using Python in a lot of my tests, particularly the WebDriver/Selenium ones, but I find the 4 space indentation and line length enforcement pretty awkward.

Maybe I'm not up-to-date on this (last checked was when writing the [`nginx-etag` test](https://github.com/NixOS/nixpkgs/blob/7ec40ece4ac19a6ae2bef2544636f1069c63a6fa/nixos/tests/nginx-etag.nix)) and it's no longer the case, but while I would prefer PEP-8 style indenting for *standalone* Python files, it's not only very weird to look at when used inline - given that Nix expressions commonly use 2 spaces for indentation - but also makes it annoying to tab twice in the editor for every block.

Same goes for the line length, I'm *theoretically* fine with the limit and I do enforce 79 chars myself when writing Nix code, but this falls short if you have things like this:

```nix
{ testScript = ''
    machine.succeed("ls ${someStorePath}")
  '';
}
```

Since `black` only knows about the *resulting* script, this would result in an error and you need to write it like this:

```nix
{ testScript = ''
    machine.succeed(
      "ls ${someStorePath}"
    )
  '';
}
```

... even though the line length is fine on the Nix side. Worse off, if you have a line where the derivation name is just long enough so `black` would break it, a simple version change of a derivation could break the validation of the test script, eg. if you have `somedrv-1.0.99` versus `somedrv-1.1.0`.

**Comment by worldofpeace** [comment]
I have the same issues https://github.com/NixOS/nixpkgs/issues/72964 @aszlig

**Comment by tfc** [comment]
> Same goes for the line length, I'm _theoretically_ fine with the limit and I do enforce 79 chars myself when writing Nix code, but this falls short if you have things like this:
> ... even though the line length is fine on the Nix side. Worse off, if you have a line where the derivation name is just long enough so `black` would break it, a simple version change of a derivation could break the validation of the test script, eg. if you have `somedrv-1.0.99` versus `somedrv-1.1.0`.

There is generally a flag to turn it off for tests. 

https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/testing-python.nix#L99

I found black relatively "patronizing" during development, but in the end it gives all python code a "uniform" look and i find that pretty appealing while reading code.
But the problem you just described sounds indeed painful. If we ever run into it in a certain frequency, i could imagine relaxing specific black rules.

**Comment by Mic92** [comment]
Some tooling would help to automatically format nixos tests using black. This is not trivial though. Especially since strings could contain interpolations from nix.

**Comment by aszlig** [comment]
@tfc: Well as mentioned, I do like having an "uniform" look, but for inline code it find the disparity between the indentation levels pretty weird to look at. For the line length issue (or maybe even the indentation issue), I think #72964 would be a better place to discuss.
